drivers/adc/adc_mcux_lpadc.c

```diff
@@ -0,0 +1,401 @@
+/*
+ * Copyright (c) 2020 Toby Firth
+ *
+ * Based on adc_mcux_adc16.c and adc_mcux_adc12.c, which are:
+ * Copyright (c) 2017-2018, NXP
+ * Copyright (c) 2019 Vestas Wind Systems A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_lpc_lpadc
+
+#include <errno.h>
+#include <drivers/adc.h>
+#include <fsl_lpadc.h>
+#include <fsl_power.h>
+
+#define LOG_LEVEL CONFIG_ADC_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(nxp_mcux_lpadc);
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+
+struct mcux_lpadc_config {
+	ADC_Type *base;
+	uint32_t clock_div;
+	uint32_t clock_source;
+	lpadc_reference_voltage_source_t voltage_ref;
+#if defined(FSL_FEATURE_LPADC_HAS_CTRL_CAL_AVGS)\
+	&& FSL_FEATURE_LPADC_HAS_CTRL_CAL_AVGS
+	lpadc_conversion_average_mode_t calibration_average;
+#endif /* FSL_FEATURE_LPADC_HAS_CTRL_CAL_AVGS */
+	lpadc_power_level_mode_t power_level;
+	uint32_t offset_a;
+	uint32_t offset_b;
+	void (*irq_config_func)(const struct device *dev);
+};
+
+struct mcux_lpadc_data {
+	const struct device *dev;
+	struct adc_context ctx;
+	uint16_t *buffer;
+	uint16_t *repeat_buffer;
+	uint32_t channels;
+	uint8_t channel_id;
+	lpadc_hardware_average_mode_t average;
+#if defined(FSL_FEATURE_LPADC_HAS_CMDL_MODE) \
+	&& FSL_FEATURE_LPADC_HAS_CMDL_MODE
+	lpadc_conversion_resolution_mode_t resolution;
+#endif /* FSL_FEATURE_LPADC_HAS_CMDL_MODE */
+};
+
+static int mcux_lpadc_channel_setup(const struct device *dev,
+				const struct adc_channel_cfg *channel_cfg)
+{
+	uint8_t channel_id = channel_cfg->channel_id;
+
+	if (channel_id > 31) {
+		LOG_ERR("Channel %d is not valid", channel_id);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		LOG_ERR("Invalid channel acquisition time");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->differential) {
+		LOG_ERR("Differential channels are not supported");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Invalid channel gain");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->reference != ADC_REF_EXTERNAL0) {
+		LOG_ERR("Invalid channel reference");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int mcux_lpadc_start_read(const struct device *dev,
+		 const struct adc_sequence *sequence)
+{
+	struct mcux_lpadc_data *data = dev->data;
+#if defined(FSL_FEATURE_LPADC_HAS_CMDL_MODE) \
+	&& FSL_FEATURE_LPADC_HAS_CMDL_MODE
+	switch (sequence->resolution) {
+	case 12:
+	case 13:
+		data->resolution = kLPADC_ConversionResolutionStandard;
+		break;
+	case 16:
+		data->resolution = kLPADC_ConversionResolutionHigh;
+		break;
+	default:
+		LOG_ERR("Unsupported resolution %d", sequence->resolution);
+		return -ENOTSUP;
+	}
+#endif /* FSL_FEATURE_LPADC_HAS_CMDL_MODE */
```

Comment:
Does it mean that if there was no `FSL_FEATURE_LPADC_HAS_CMDL_MODE`, any requested resolution would be supported?

Reply:
I have rechecked the documentation and if FSL_FEATURE_LPADC_HAS_CMDL_MODE is not enabled then there is no resolution configuration option avaible for the ADC. Which in effect means that 12/13 bit resolution in the only option. So I need to add an #else clause to the #ifdef which checks that sequence->resolution is 12/13.